### PR TITLE
Implement support for JSX syntax.

### LINKED ISSUE #1467: Support JSX Syntax
Optionally, Esprima should be able to consume JSX syntax: https://facebook.github.io/jsx/. See also https://github.com/facebook/jsx/blob/master/AST.md.

The grammar extension is summarized below (adopted from the specification linked above).

##### Lexical Grammar

*JSXIdentifier :: JSXIdentifierStart | JSXIdentifier JSXIdentifierPart*

*JSXIdentifierStart :: IdentifierStart (__but not__ `\`)*

*JSXIdentifierPart :: IdentifierPart (__but_not__ `\`) | `-`*

*JSXText :: JSXTextCharacter JSXText<sub>opt</sub>*

*JSXTextCharacter :: SourceCharacter __but not one of__ `{`, `<`, `>` or `}`*

*JSXStringLiteral :: `'` JSXText `'` | `"` JSXText `"`*

##### Primary expression

JSX extends the PrimaryExpression in the ECMAScript 6th Edition (ECMA-262) grammar:

*PrimaryExpression :: JSXElement*

##### Elements

Note: names of JSXOpeningElement and JSXClosingElement should match.

*JSXElement :: JSXSelfClosingElement | JSXOpeningElement JSXChildren<sub>opt</sub> JSXClosingElement*

*JSXSelfClosingElement :: `<` JSXElementName JSXAttributes<sub>opt</sub> `/` `>`*

*JSXOpeningElement :: `<` JSXElementName JSXAttributes<sub>opt</sub> `>`*

*JSXClosingElement :: `<` `/` JSXElementName `>`*

*JSXElementName :: JSXIdentifier | JSXNamedspacedName | JSXMemberExpression*

*JSXNamespacedName :: JSXIdentifier `:` JSXIdentifier*

*JSXMemberExpression :: JSXIdentifier `.` JSXIdentifier | JSXMemberExpression `.` JSXIdentifier*

##### Attributes

*JSXAttributes :: JSXSpreadAttribute JSXAttributes<sub>opt</sub> | JSXAttribute JSXAttributes<sub>opt</sub>*

*JSXSpreadAttribute :: `{` `...` AssignmentExpression `}`*

*JSXAttribute :: JSXAttributeName | JSXAttributeName `=` JSXAttributeValue*

*JSXAttributeName :: JSXIdentifier | JSXNamespacedName*

*JSXAttributeValue :: StringLiteral | `{` AssignmentExpression `}` | JSXElement*

##### Children

*JSXChildren : JSXChild JSXChildren<sub>opt</sub>*

*JSXChild :: JSXText | JSXElement | `{` AssignmentExpression<sub>opt</sub> `}`*

### PR DESCRIPTION
**Pull request by ariya**
Example:

```js
  esprima.parse('<div>Hello</div>', { jsx: true });
```
Fixes #1467

### CODE REVIEW

File: src/xhtml-entities.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    rceil: '\u2309',
+    lfloor: '\u230A',
+    rfloor: '\u230B',
+    loz: '\u25CA',
+    spades: '\u2660',
+    clubs: '\u2663',
+    hearts: '\u2665',
+    diams: '\u2666',
+    lang: '\u27E8',
+    rang: '\u27E9'
```
Comment (ariya) [inline src/xhtml-entities.ts]: This is different than the old esprima-fb, which uses U+2329 for `lang` and U+232A for `rang` (I'm not sure why).

If we use https://www.w3.org/TR/xml-entity-names/xhtml1-symbol.html as a reference, it clearly says that `lang` is U+27E8 and `rang` is U+27E9.

File: src/jsx-parser.ts [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+                children.push(container);
+            } else {
+                break;
+            }
+        }
+
+        return children;
+    }
+
+    parseComplexJSXElement(el: MetaJSXElement): MetaJSXElement {
```
Comment (ariya) [inline src/jsx-parser.ts]: This is a stack-based implementation, different than the recursive variant in the old esprima-fb. It eliminates the need for the double lookahead, and yet it should still conform to the specification.

File: test/fixtures/JSX/empty-expression-container.js [hunk 3]
```diff
@@ -0,0 +1 @@
+<body>{}</body>
```
Comment (ariya) [inline test/fixtures/JSX/empty-expression-container.js]: This test demonstrates an inconsistency in Flow (which is not followed in this pull request).

Consider the following two cases, `<a>{ }</a>` and `<a>{b}</a>`, and what Flow produces. In the first case, the location range for the inner JSXEmptyExpression node includes the curly braces, i.e. the same as the location range of the enclosing JSXExpressionContainer node. In the second case, the Identifier node for `b` obviously does not include the curly braces.

**Comment by ariya** [comment]
@sebmarkbage @jeffmo @mroch Care to review?

**Comment by mikesherov** [comment]
I'm not qualified to review, but.... :+1:

**Pull request closed without merging**